### PR TITLE
feat(module: tree): add ExpandAll/CollapseAll TreeNode

### PR DESCRIPTION
**Pull request by ardia**
expand/collapse node and expand/collapse all child nodes

https://github.com/ant-design-blazor/ant-design-blazor/issues/2370

### 🤔 This is a ...

- [x] New feature
- [ ] Bug fix
- [ ] Site / documentation update
- [ ] Demo update
- [ ] Component style update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [ ] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Other (about what?)

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English |           |
| 🇨🇳 Chinese |           |

### ☑️ Self Check before Merge

⚠️ Please check all items below before review. ⚠️

- [ ] Doc is updated/provided or not needed
- [x] Demo is updated/provided or not needed
- [ ] Changelog is provided or not needed

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/ant-design-blazor/ant-design-blazor?pullRequest=2395) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/ant-design-blazor/ant-design-blazor?pullRequest=2395) before we can accept your contribution.<br/><br/>:x: ardia [sign now](https://cla.dotnetfoundation.org/ant-design-blazor/ant-design-blazor?pullRequest=2395)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/ant-design-blazor/ant-design-blazor?pullRequest=2395) it.</sub>

File: components/tree/TreeNode.razor.cs [hunk 1]
```diff
@@ -343,6 +343,33 @@ public void Expand(bool expanded)
...
+        {
+            Switch(this, false);
+        }
+
+        /// <summary>
+        /// 节点展开关闭
+        /// </summary>
+        /// <param name="node"></param>
+        /// <param name="expanded"></param>
+        private void Switch(TreeNode<TItem> node, bool expanded)
```
Comment (ElderJames) [inline components/tree/TreeNode.razor.cs:367]: I think it's better called `SwitchAllNodes`.

**Comment by ElderJames** [comment]
Have you considered asynchronous loading?

**Comment by ardia** [comment]
> Have you considered asynchronous loading?

This doesn't work. Expand all tree also doesn't work.

load data and expand

`private void ExpandAll()
    {
        _tree.ExpandAll();
    }   
`

`private void ExpandAllChild()
    {
        _tree.FindFirstOrDefaultNode(n => n.Title == "C").ExpandAll();
    }`

**Comment by ElderJames** [comment]
Yeah, so we need to think about how to solve it. Such as loading child nodes when expanding.

**Comment by ardia** [comment]
> Yeah, so we need to think about how to solve it. Such as loading child nodes when expanding.

expand async load

File: components/tree/Tree.razor.cs [hunk 2]
```diff
@@ -687,26 +687,26 @@ public TreeNode<TItem> FindFirstOrDefaultNode(Func<TreeNode<TItem>, bool> predic
...
+            ChildNodes.ForEach(node => _ = Switch(node, false));
         }
 
         /// <summary>
         /// 节点展开关闭
         /// </summary>
         /// <param name="node"></param>
         /// <param name="expanded"></param>
-        private void Switch(TreeNode<TItem> node, bool expanded)
+        private async Task Switch(TreeNode<TItem> node, bool expanded)
```
Comment (ElderJames) [inline components/tree/Tree.razor.cs:706]: Are the `Switch` and `SwitchAllNodes` duplicated?

Comment (ardia) [reply components/tree/Tree.razor.cs]: Yes, that's fixed.

File: components/tree/TreeNode.razor.cs [hunk 3]
```diff
@@ -337,10 +337,39 @@ public bool IsLeaf
...
             if (Expanded == expanded) return;
             Expanded = expanded;
+
+            await TreeComponent?.OnNodeExpand(this, Expanded, new MouseEventArgs());
+        }
+
+        /// <summary>
+        /// Expand all child nodes
+        /// </summary>
+        public async Task ExpandAll()
```
Comment (ElderJames) [inline components/tree/TreeNode.razor.cs:351]:
```suggestion
        internal async Task ExpandAll()
```

File: components/tree/TreeNode.razor.cs [hunk 4]
```diff
@@ -337,10 +337,39 @@ public bool IsLeaf
...
+        {
+            await SwitchAllNodes(this, true);
+        }
+
+        /// <summary>
+        /// Collapse all child nodes
+        /// </summary>
+        public async Task CollapseAll()
```
Comment (ElderJames) [inline components/tree/TreeNode.razor.cs:359]:
```suggestion
        internal async Task CollapseAll()
```

File: components/tree/TreeNode.razor.cs (hunk 3 above)
Comment (ardia) [reply components/tree/TreeNode.razor.cs:351]: if to internal
`_tree.FindFirstOrDefaultNode(n => n.Title == "C").ExpandAll();`
not work

File: components/tree/TreeNode.razor.cs (hunk 4 above)
Comment (ardia) [reply components/tree/TreeNode.razor.cs:359]: if to internal
_tree.FindFirstOrDefaultNode(n => n.Title == "C").CollapseAll();
not work

Comment (ElderJames) [reply components/tree/TreeNode.razor.cs:359]: If you need to expand a node individually, I think you should set up a cleaner API.
How about：
```
_tree.ExpandAll("node-key")
```

**Comment by ElderJames** [comment]
@ardia Sorry, I cleaned the warehouse, so this PR was closed. Please rebase to the master branch and re-open the PR.

**Comment by ElderJames** [comment]
Hi @ardia ,could you please create another new PR base on feature branch?

**Pull request closed without merging**